src/adapter/src/catalog.rs

```diff
@@ -1395,18 +1395,62 @@ impl ExprHumanizer for ConnCatalog<'_> {
     }
 
     fn column_names_for_id(&self, id: GlobalId) -> Option<Vec<String>> {
+        let Some(entry) = self.state.entry_by_id.get(&id) else {
+            return None;
+        };
+
+        match entry.index() {
+            Some(index) => {
+                let Some(on_entry) = self.state.entry_by_id.get(&index.on) else {
+                    return None;
+                };
+                let Ok(on_desc) = on_entry.desc(&self.resolve_full_name(on_entry.name())) else {
+                    return None;
+                };
+
+                let mut on_names = on_desc
+                    .iter_names()
+                    .map(|col_name| col_name.to_string())
+                    .collect::<Vec<_>>();
+
+                let (p, _) = mz_expr::permutation_for_arrangement(&index.keys, on_desc.arity());
+
+                // Init ix_names with unknown column names.
+                let ix_arity = p.keys().cloned().max().expect("max key") + 1;
+                let mut ix_names = vec![String::new(); ix_arity];
+
+                // Apply the permutation by swapping on_names with ix_names.
+                for (on_pos, ix_pos) in p.into_iter() {
+                    let on_name = on_names.get_mut(on_pos).expect("on_name");
+                    let ix_name = ix_names.get_mut(ix_pos).expect("ix_name");
+                    std::mem::swap(on_name, ix_name);
+                }
+
+                Some(ix_names) // Return the updated ix_names vector.
+            }
```

Comment:
Not covered by any test case (if the [coverage data](https://buildkite.com/materialize/coverage/builds/408) is accurate)?

Reply:
This cannot be correct—for example queries that generate optimizer notices rely on this code.